### PR TITLE
Initial sketch of logging API usage docs.

### PR DESCRIPTION
**Pull request by tseaver**
@jgeewax @dhermes please review the usage patterns here.  Note that I've left out the API methods for listing `monitoredResourceDescriptors`:  I couldn't quite figure why they are part of this API, rather than a more generic one (such as `resource_manager`).

I don't intend to merge this PR directly, but will incorporate the updated usage doc in the initial PR for logging.  BTW, should we adopt the pattern of targeting such PRs at a long-lived `logging-api` branch, rather than master?  We would then land the whole API at once by merging that branch.

### CODE REVIEW

**Comment by dhermes** [comment]
@tseaver Hopefully [this][1] will resolve the pull request builds for AppVeyor. Was

screen_shot_002

and now is

screen_shot_003


[1]: http://help.appveyor.com/discussions/suggestions/373-disable-build-for-pullrequest

**Comment by dhermes** [comment]
RE: `logging-api` branch, I'm "fine" with it, though gardening the merge base is inherently painful and it makes the "final review" more difficult as well when it gets merged back into `master`.

**Comment by tseaver** [comment]
Given that I don't see changes on the putative 'logging-api' branch touching any common code, the final merge should be pretty trivial.

**Comment by dhermes** [comment]
Yeah it seems less painful than the impending `datastore-v1beta3` merge.

File: docs/logging-usage.rst [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+  your applications.
+
+- :class:`Client <gcloud.logging.client.Client>` objects hold both a ``project``
+  and an authenticated connection to the Logging service.
+
+- The authentication credentials can be implicitly determined from the
+  environment or directly via
+  :meth:`from_service_account_json <gcloud.logging.client.Client.from_service_account_json>`
+  and
+  :meth:`from_service_account_p12 <gcloud.logging.client.Client.from_service_account_p12>`.
```
Comment (dhermes) [inline docs/logging-usage.rst]: I realized in #1420 that this could just be a reference to

```
:doc:`gcloud-auth`
```

File: docs/logging-usage.rst [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+Writing log entries
+-------------------
+
+Write a simple text entry to a log.
+
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> log = client.log('log_name')
```
Comment (dhermes) [inline docs/logging-usage.rst]: I worry here that `log` the noun and `log` the verb will be confused.

File: docs/logging-usage.rst [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+   >>> log.text("A simple entry")  # API call
```
Comment (dhermes) [inline docs/logging-usage.rst]: `text` isn't a very clear name for an API call. What does it do?

File: docs/logging-usage.rst [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+
+Write a dictionary entry to a log.
+
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> log = client.log('log_name')
+   >>> log.struct(message="My second entry",
+   ...     weather="partly cloudy")  # API call
```
Comment (dhermes) [inline docs/logging-usage.rst]: Dan the human linter detects a malfunction

File: docs/logging-usage.rst [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+   >>> log.text("A simple entry")  # API call
+
+Write a dictionary entry to a log.
+
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> log = client.log('log_name')
+   >>> log.struct(message="My second entry",
```
Comment (dhermes) [inline docs/logging-usage.rst]: Ditto here on the verb (`struct`, like `text` above) not feeling right

File: docs/logging-usage.rst [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+
+- After setting ``GOOGLE_APPLICATION_CREDENTIALS`` and ``GCLOUD_PROJECT``
+  environment variables, create a :class:`Client <gcloud.logging.client.Client>`
+
+  .. doctest::
+
+     >>> from gcloud import logging
+     >>> client = logging.Client()
+
+Writing log entries
```
Comment (dhermes) [inline docs/logging-usage.rst]: Safe to assume both methods below use [`logging.entries.write`][1]?

[1]: https://developers.google.com/apis-explorer/#p/logging/v2beta1/logging.entries.write

File: docs/logging-usage.rst [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> entries, token = client.list_entries()  # API call
+   >>> for entry in entries:
+   ...    timestamp = entry.timestamp.ISO()
+   ...    print(timestamp, entry.text_payload, entry.struct_payload)
+   ('2016-02-17T20:35:49.031864072Z', 'A simple entry', None)
+   ('2016-02-17T20:38:15.944418531Z,' None, {'message': 'My second entry', 'weather': 'partly cloudy'})
```
Comment (dhermes) [inline docs/logging-usage.rst]: More human linter fun

```python
>>> # py2
>>> print(1, 2, 3)
(1, 2, 3)
>>> # py3
>>> print(1, 2, 3)
1 2 3
```

File: docs/logging-usage.rst [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+
+Fetch entries for the default project.
+
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> entries, token = client.list_entries()  # API call
+   >>> for entry in entries:
+   ...    timestamp = entry.timestamp.ISO()
```
Comment (dhermes) [inline docs/logging-usage.rst]: What type implements `ISO`?

File: docs/logging-usage.rst [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+Retrieving log entries
+----------------------
+
+Fetch entries for the default project.
+
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> entries, token = client.list_entries()  # API call
```
Comment (dhermes) [inline docs/logging-usage.rst]: When do we decide to use an `__iter__`-able versus here? (Which is more like `client.get_entries_page()` or something like that.)

File: docs/logging-usage.rst [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+Fetch entries across multiple projects.
+
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> entries, token = client.list_entries(
+   ...     project_ids=['one-project', 'another-project'])  # API call
+
+
```
Comment (dhermes) [inline docs/logging-usage.rst]: How do you decide between 1 or 2 spaces between these sections. Unclear which one you were going for.

File: docs/logging-usage.rst [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+Filter entries retrieved using the "Advance Logs Filters" syntax (see
+https://cloud.google.com/logging/docs/view/advanced_filters).
```
Comment (dhermes) [inline docs/logging-usage.rst]: Sphinx style nit: Use a link instead of inline URL?

File: docs/logging-usage.rst [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> FILTER = "log:log_name AND textPayload:simple"
```
Comment (dhermes) [inline docs/logging-usage.rst]: How complex are the filters? Not worth having our own native type to build them?

File: docs/logging-usage.rst [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+   >>> entries, token = client.list_entries(filter=FILTER)  # API call
+
+Sort entries in descending timestamp order.
+
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> entries, token = client.list_entries(order_by='timestamp desc')  # API call
```
Comment (dhermes) [inline docs/logging-usage.rst]: How does `timestamp desc` get passed to the API?

Comment (dhermes) [inline docs/logging-usage.rst]: Are multiple orders possible? First `A`, then `B`, then `C`, etc.?

File: docs/logging-usage.rst [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+used within Cloud Monitoring to create charts and alerts.
+
+Create a metric:
+
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> metric = client.metric("robots", "Robots all up in your server",
+   ...     filter='log:apache-access AND textPayload:robot')
```
Comment (dhermes) [inline docs/logging-usage.rst]: Human linter says indent

File: docs/logging-usage.rst [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+List all metrics for a project:
+
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> metrics, token = client.list_metrics()
+   >>> len(metrics)
+   0
+   >>> metric = metrics[0]
```
Comment (dhermes) [inline docs/logging-usage.rst]: If it's empty then how does `metrics[0]` work?

File: docs/logging-usage.rst [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+Update a metric:
+
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> metric = client.metric("robots")
+   >>> metric.exists()  # API call
+   True
+   >>> metric.get()  # API call
```
Comment (dhermes) [inline docs/logging-usage.rst]: Seems to be the same thing we use `reload()` for elsewhere?

File: docs/logging-usage.rst [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+   >>> metric.delete()  # API call
+   >>> metric.exists()  # API call
+   False
+
+
+Export log entries using sinks
+------------------------------
+
+Sinks allow exporting entries which match a given filter to Cloud Storage
+buckets, BigQuery datasets, or Pubsub topics.
```
Comment (dhermes) [inline docs/logging-usage.rst]: s/`Pubsub`/`Pub/Sub`

File: docs/logging-usage.rst [hunk 18]
```diff
@@ -0,0 +1,295 @@
...
+
+Create a CloudStorage sink:
```
Comment (dhermes) [inline docs/logging-usage.rst]: Any reason you used CloudStorage?

File: docs/logging-usage.rst [hunk 19]
```diff
@@ -0,0 +1,295 @@
...
+
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> sink = client.sink("robots-storage",
+   ...     filter='log:apache-access AND textPayload:robot')
```
Comment (dhermes) [inline docs/logging-usage.rst]: Human linter says indent

File: docs/logging-usage.rst [hunk 20]
```diff
@@ -0,0 +1,295 @@
...
+   True
+
+Create a BigQuery sink:
+
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> sink = client.sink("robots-bq",
+   ...     filter='log:apache-access AND textPayload:robot')
```
Comment (dhermes) [inline docs/logging-usage.rst]: Human linter

File: docs/logging-usage.rst [hunk 21]
```diff
@@ -0,0 +1,295 @@
...
+   >>> sink.bigquery_dataset = "projects/my-project/datasets/my-dataset"
```
Comment (dhermes) [inline docs/logging-usage.rst]: Is the `bigquery_` qualifier necessary? Would just `dataset` suffice?

File: docs/logging-usage.rst [hunk 22]
```diff
@@ -0,0 +1,295 @@
...
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> sinks, token = client.list_sinks()
+   >>> for sink in sinks:
+   ...     print(sink.name, sink.destination)
+   ('robots-storage', 'storage.googleapis.com/my-bucket-name')
+   ('robots-bq', 'bigquery.googleapis.com/projects/my-project/datasets/my-dataset')
+   ('robots-pubsub', 'pubsub.googleapis.com/projects/my-project/topics/my-topic')
```
Comment (dhermes) [inline docs/logging-usage.rst]: Ditto here on the output differing btw. py2 and py3

File: docs/logging-usage.rst [hunk 23]
```diff
@@ -0,0 +1,295 @@
...
+
+List all sinks for a project:
+
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> sinks, token = client.list_sinks()
+   >>> for sink in sinks:
+   ...     print(sink.name, sink.destination)
```
Comment (dhermes) [inline docs/logging-usage.rst]: Destination seems to be free-form (a bucket, a dataset, a topic, a man, a plan, a canal panama)

File: docs/logging-usage.rst [hunk 24]
```diff
@@ -0,0 +1,295 @@
...
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> sink = client.sink('robots-storage')
+   >>> sink.filter is None
+   True
+   >>> sink.get()  # API call
+   >>> sink.filter
+   'log:apache-access AND textPayload:robot'
+   >>> sink.destination
+   'storage.googleapis.com/my-bucket-name')
```
Comment (dhermes) [inline docs/logging-usage.rst]: Trailing `)`

File: docs/logging-usage.rst [hunk 25]
```diff
@@ -0,0 +1,295 @@
...
+Refresh local information about a sink:
+
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> sink = client.sink('robots-storage')
+   >>> sink.filter is None
+   True
+   >>> sink.get()  # API call
```
Comment (dhermes) [inline docs/logging-usage.rst]: Ditto here on `get` vs. `reload`

File: docs/logging-usage.rst [hunk 26]
```diff
@@ -0,0 +1,295 @@
...
+   >>> sink.update()  # API call
+
+Delete a sink:
+
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> sink = client.sink("robots",
+   ...     filter='log:apache-access AND textPayload:robot')
```
Comment (dhermes) [inline docs/logging-usage.rst]: Human linter

File: docs/logging-usage.rst (hunk 3 above)
Comment (tseaver) [inline docs/logging-usage.rst]: I originally considered `add_text_entry` (vs. `add_struct_entry` below), but they seemed too verbose.  WDYT?

File: docs/logging-usage.rst (hunk 4 above)
Comment (tseaver) [inline docs/logging-usage.rst]: I'm afraid I don't see it:  can you be more precise? :)

File: docs/logging-usage.rst (hunk 9 above)
Comment (tseaver) [inline docs/logging-usage.rst]: We have other APIs (e.g., `pubsub.client.Client.list_topics`,  `pubsub.client.Client.list_subscriptions`, `bigquery.client.Client.list_datasets`) which follow that pattern.

The problem with an `__iter__`-based solution is that it necessarily hides when the API calls occur.

File: docs/logging-usage.rst (hunk 8 above)
Comment (tseaver) [inline docs/logging-usage.rst]: `datetime.datetime`.

File: docs/logging-usage.rst (hunk 7 above)
Comment (tseaver) [inline docs/logging-usage.rst]: I guess we could go for a format string then (old-skool!)

File: docs/logging-usage.rst (hunk 10 above)
Comment (tseaver) [inline docs/logging-usage.rst]: Should be one:  I'm only likely to use two between major sections.

File: docs/logging-usage.rst (hunk 12 above)
Comment (tseaver) [inline docs/logging-usage.rst]: They are arbitrary boolean expressions (something like the `WHERE` clause in SQL).  I would definitely class a builder for them as being post-MVP.

File: docs/logging-usage.rst (hunk 13 above)
Comment (tseaver) [inline docs/logging-usage.rst]: Currently, the only allowed values are `timestamp asc` and `timestamp desc`:  the [spec](https://cloud.google.com/logging/docs/api/ref_v2beta1/rest/v2beta1/entries/list#body.request_body.FIELDS.order_by) doesn't seem to promise any future variations.

File: docs/logging-usage.rst (hunk 14 above)
Comment (tseaver) [inline docs/logging-usage.rst]: You need to tell me how much:  I already indented the four spaces that seemed right to me.

File: docs/logging-usage.rst (hunk 15 above)
Comment (tseaver) [inline docs/logging-usage.rst]: `len(metrics)` should be 1.

File: docs/logging-usage.rst (hunk 16 above)
Comment (tseaver) [inline docs/logging-usage.rst]: Agreed:  I should have looked more closely.

File: docs/logging-usage.rst (hunk 6 above)
Comment (tseaver) [inline docs/logging-usage.rst]: Yes.

File: docs/logging-usage.rst (hunk 2 above)
Comment (tseaver) [inline docs/logging-usage.rst]: I'm conflicted:  the API [refers to the resource type as `logs`](https://cloud.google.com/logging/docs/api/ref_v2beta1/rest/v2beta1/projects.logs), but the more natural name in Python land might be "logger" (except, of course, that it is not related remotely to the stdlib's `logger`).

File: docs/logging-usage.rst (hunk 18 above)
Comment (tseaver) [inline docs/logging-usage.rst]: I was looking at the [logging docs for export](https://cloud.google.com/logging/docs/api/tasks/exporting-logs), which use `Cloud Storage` (they also use `Cloud Pub/Sub`).

File: docs/logging-usage.rst (hunk 1 above)
Comment (tseaver) [inline docs/logging-usage.rst]: f413435

File: docs/logging-usage.rst (hunk 8 above)
Comment (tseaver) [inline docs/logging-usage.rst]: Oops, my bad.  4d61743

File: docs/logging-usage.rst (hunk 7 above)
Comment (tseaver) [inline docs/logging-usage.rst]: 4d61743

File: docs/logging-usage.rst (hunk 10 above)
Comment (tseaver) [inline docs/logging-usage.rst]: 0deb356

File: docs/logging-usage.rst (hunk 11 above)
Comment (tseaver) [inline docs/logging-usage.rst]: 3bf1e7a

File: docs/logging-usage.rst (hunk 15 above)
Comment (tseaver) [inline docs/logging-usage.rst]: 97c2874

File: docs/logging-usage.rst (hunk 16 above)
Comment (tseaver) [inline docs/logging-usage.rst]: 15bdcb6

File: docs/logging-usage.rst (hunk 17 above)
Comment (tseaver) [inline docs/logging-usage.rst]: 4bbdd17

File: docs/logging-usage.rst (hunk 18 above)
Comment (tseaver) [inline docs/logging-usage.rst]: 885c134

File: docs/logging-usage.rst (hunk 21 above)
Comment (tseaver) [inline docs/logging-usage.rst]: `dataset` might be ambiguous / confusing (`datastore` uses it, too).

**Comment by tseaver** [comment]
@dhermes PTAL:  I will squash all the fixup commits when merging to the `logging-api` branch.

Also, can you please point me to an example of how you think the docstring continuation lines should be indented?

File: docs/logging-usage.rst (hunk 22 above)
Comment (tseaver) [inline docs/logging-usage.rst]: 4d61743a6061be93943f53bb8158d4a7785adb2e

File: docs/logging-usage.rst (hunk 23 above)
Comment (tseaver) [inline docs/logging-usage.rst]: The possible URL formats for the three services are[specified in the [logging export docs](https://cloud.google.com/logging/docs/api/tasks/exporting-logs#about_sinks).

File: docs/logging-usage.rst (hunk 24 above)
Comment (tseaver) [inline docs/logging-usage.rst]: 22cbb49

File: docs/logging-usage.rst (hunk 25 above)
Comment (tseaver) [inline docs/logging-usage.rst]: 15bdcb6a95174312d8f02066d38292bdd8545a53

File: docs/logging-usage.rst (hunk 4 above)
Comment (tseaver) [inline docs/logging-usage.rst]: I went looking to find the pattern you use for doctest continuations, and found [this example](https://github.com/GoogleCloudPlatform/gcloud-python/pull/1420/files#diff-50dad309c06e8edd83f86b8e40b32a85R115), which shows:

```python
   >>> bucket1, bucket2, bucket3 = client.get_buckets(
   ...     'bucket-name1',
   ...     'bucket-name2',
   ...     'bucket-name3')  # API request
```

i.e., four additional spaces after the `...`.

Comment (dhermes) [inline docs/logging-usage.rst]: Key difference is what comes before the `(`.

Option 1:

```python
>>> log.struct(message="My second entry",
...            weather="partly cloudy")  # API call
```

Option 2

```python
>>> log.struct(
...     message="My second entry", weather="partly cloudy")  # API call
```

Option 3

```python
>>> log.struct(
...     message="My second entry",
...     weather="partly cloudy")  # API call
```

File: docs/logging-usage.rst (hunk 3 above)
Comment (dhermes) [inline docs/logging-usage.rst]: I think they are a tad too verbose but are still superior to `text`/`struct` as alternatives.

File: docs/logging-usage.rst (hunk 2 above)
Comment (dhermes) [inline docs/logging-usage.rst]: Yeah it's a very overloaded term. We could just make one up? We have precedent for this, `object` -> `blob` in storage because Python.

File: docs/logging-usage.rst (hunk 9 above)
Comment (dhermes) [inline docs/logging-usage.rst]: I'm equally worried about the

> When do we decide

as I am about what the answer is. I mostly agree with you, though I think there may be a subset of users that wants paging hidden from them.

Most importantly, we should be consistent across services.

File: docs/logging-usage.rst (hunk 13 above)
Comment (dhermes) [inline docs/logging-usage.rst]: Ha OK. Might be worth providing them as `Client.ASCENDING` and `Client.DESCENDING`?

File: docs/logging-usage.rst (hunk 14 above)
Comment (dhermes) [inline docs/logging-usage.rst]: See above.

File: docs/logging-usage.rst (hunk 4 above)
Comment (dhermes) [inline docs/logging-usage.rst]: FWIW the [indentation][1] section of PEP8 is what I'm going with here (not me being arbitrary). AFAICT, there isn't much room for interpretation.

[1]: https://www.python.org/dev/peps/pep-0008/#indentation

File: docs/logging-usage.rst (hunk 21 above)
Comment (dhermes) [inline docs/logging-usage.rst]: OK. (Though as of `v1beta3` the `dataset` term is gonezo.)

File: docs/logging-usage.rst [hunk 27]
```diff
@@ -0,0 +1,301 @@
...
+Fetch entries for the default project.
+
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> entries, token = client.list_entries()  # API call
+   >>> for entry in entries:
+   ...    timestamp = entry.timestamp.isoformat()
+   ...    print('%sZ: %s | %s" %
```
Comment (dhermes) [inline docs/logging-usage.rst]: Single quote at start, double at end.

File: docs/logging-usage.rst [hunk 28]
```diff
@@ -0,0 +1,301 @@
...
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> entries, token = client.list_entries(
+   ...     project_ids=['one-project', 'another-project'])  # API call
+
+Filter entries retrieved using the `Advanced Logs Filters`_ syntax
+
+.. Advanced Logs Filters: https://cloud.google.com/logging/docs/view/advanced_filters
```
Comment (dhermes) [inline docs/logging-usage.rst]: Sorry I messed up the syntax in my suggestion. Should be

```
.. _Advanced Logs Filters: https://cloud.google.com/logging/docs/view/advanced_filters
```

File: docs/logging-usage.rst (hunk 2 above)
Comment (tseaver) [inline docs/logging-usage.rst]: 276b27b

File: docs/logging-usage.rst (hunk 3 above)
Comment (tseaver) [inline docs/logging-usage.rst]: 93a7e31

File: docs/logging-usage.rst (hunk 5 above)
Comment (tseaver) [inline docs/logging-usage.rst]: 93a7e31

File: docs/logging-usage.rst (hunk 26 above)
Comment (tseaver) [inline docs/logging-usage.rst]: 449e4f0

File: docs/logging-usage.rst (hunk 27 above)
Comment (tseaver) [inline docs/logging-usage.rst]: bf5b5cc

File: docs/logging-usage.rst (hunk 28 above)
Comment (tseaver) [inline docs/logging-usage.rst]: e18acc7

File: docs/logging-usage.rst (hunk 13 above)
Comment (tseaver) [inline docs/logging-usage.rst]: 4a3bdae

File: docs/logging-usage.rst (hunk 9 above)
Comment (tseaver) [inline docs/logging-usage.rst]: We got feedback from the service teams that "implicit" service calls were a Bad Thing (TM).

We could probably write a generic iterator factory which would wrap around the "paged" methods, taking extra arguments as `**kw` to the factory.  E.g.:

```python
from gcloud.iterator import paging_iterator

for item in paging_iterator(client.list_entries, filter=FILTER, order_by=Client.DESCENDING):
    frob(item)
```

As long as the return values are the same, i.e., `(page_items, next_token)`, and the argument name for passing the previous token back in is consistent, it should Just Work (TM).

Comment (dhermes) [inline docs/logging-usage.rst]: I don't recall that feedback. The way I remember it, that was **our own** feedback? Hidden behavior has caused funky errors like #1467.

Comment (tseaver) [inline docs/logging-usage.rst]: Hmm,  my memory of the discussion involves the initial feedback @tmatsuo gave us on the pubsub usage sketch:  he was keen both on removing "required-but-useless" API requests, and on making it very explicit which calls would trigger one.  ISTM that log entries are the perfect example for this:  unlike pubsub subscriptions / topics, bigquery datasets / tables, etc., a large project is likely to have an arbitrarily-large number of log entries over time.  Accidentally iterating over all of them by accident could be hideously expensive (both in runtime performance and in cost).

Comment (dhermes) [inline docs/logging-usage.rst]: Yeah I agree. Just want to focus on "Most importantly, we should be consistent across services."

**Comment by tseaver** [comment]
@dhermes any remaining issues?  If not, can you assign to @jgeewax for review of the API surface?

File: docs/logging-usage.rst [hunk 29]
```diff
@@ -0,0 +1,307 @@
...
+-------------------
+
+Write a simple text entry to a logger.
+
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> logger = client.logger('log_name')
+   >>> logger.log_text("A simple entry")  # API call
```
Comment (jgeewax) [inline docs/logging-usage.rst]: Is there a way to automagically figure this out? So that I could write:

```python
logger.log('Some text!')
logger.log(message='this is some data', severity='ERROR')
logger.log(exception)
```

?

Comment (dhermes) [inline docs/logging-usage.rst]: @tseaver Filed #1497 to discuss just such a pattern

File: docs/logging-usage.rst (hunk 9 above)
Comment (jgeewax) [inline docs/logging-usage.rst]: I recall participating in this argument a few times before, still hating the fact that we expose a token to people, who then go "WTF? I wanted to just list all the things... why are you giving me a token and making *me* write code? *you* should write it!"

If users want to be explicit about paging -- which they sometimes will -- that should be possible, but I don't love that as the default.

In other words, listing `thing`s:

```python
# Just give me all the things.
for thing in client.list_things(some=extra, params=here):
  print thing.name

# I actually want to paginate through these, because I'm building a web app that displays them:
page_token = request.get('page_token')
things, next_token = client.get_things_page(page_token=page_token, page_size=20, some=extra, params=here)
return response.render_template('my_template.html', context={'things': things})
```

I totally see the value in having a token and paginating through, but I think that doesn't follow our mantra of "make the common-case fast (to write)".

File: docs/logging-usage.rst (hunk 12 above)
Comment (jgeewax) [inline docs/logging-usage.rst]: +1 , post-MVP sounds like the right time to dig into that.

File: docs/logging-usage.rst [hunk 30]
```diff
@@ -0,0 +1,307 @@
...
+   >>> FILTER = "log:log_name AND textPayload:simple"
+   >>> entries, token = client.list_entries(filter=FILTER)  # API call
+
+Sort entries in descending timestamp order.
+
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> entries, token = client.list_entries(order_by=Client.DESCENDING)  # API call
```
Comment (jgeewax) [inline docs/logging-usage.rst]: A bit confused by `Client.DESCENDING`. Couldn't this be `logging.DESCENDING` ?

File: docs/logging-usage.rst [hunk 31]
```diff
@@ -0,0 +1,307 @@
...
+Retrieve entities in batches of 10, iterating until done.
+
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> retrieved = []
+   >>> token = None
+   >>> while True:
+   ...     entries, token = client.list_entries(page_size=10)  # API call
```
Comment (jgeewax) [inline docs/logging-usage.rst]: Not sure I understand how this works. Don't you need to tell it about the next token? If not, then why return it at all ? We could just look while `len(entries) > 0`

File: docs/logging-usage.rst [hunk 32]
```diff
@@ -0,0 +1,307 @@
...
+
+Delete all entries for a logger
+-------------------------------
+
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> logger = client.logger('log_name')
+   >>> logger.delete()  # API call
```
Comment (jgeewax) [inline docs/logging-usage.rst]: Hmm... this scares me. Not sure I expect deleting a logger to delete all the entries I've stored over a long while. Maybe `.delete_entries()` ?

File: docs/logging-usage.rst [hunk 33]
```diff
@@ -0,0 +1,307 @@
...
+Create a BigQuery sink:
+
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> sink = client.sink(
+   ...     "robots-bq",
+   ...     filter='log:apache-access AND textPayload:robot')
+   >>> sink.bigquery_dataset = "projects/my-project/datasets/my-dataset"
```
Comment (jgeewax) [inline docs/logging-usage.rst]: We can't set this in the initializer ?

File: docs/logging-usage.rst [hunk 34]
```diff
@@ -0,0 +1,307 @@
...
+Create a Cloud Pub/Sub sink:
+
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> sink = client.sink(
+   ...     "robots-pubsub",
+   ...     filter='log:apache-access AND textPayload:robot')
+   >>> sink.pubsub_topic = 'projects/my-project/topics/my-topic'
```
Comment (jgeewax) [inline docs/logging-usage.rst]: What happens if we set multiple of these ? pubsub_topic *and* bigquery_dataset ?

File: docs/logging-usage.rst [hunk 35]
```diff
@@ -0,0 +1,307 @@
...
+List all sinks for a project:
+
+.. doctest::
+
+   >>> from gcloud import logging
+   >>> client = logging.Client()
+   >>> sinks, token = client.list_sinks()
+   >>> for sink in sinks:
+   ...     print('%s: %s' % (sink.name, sink.destination))
+   robots-storage: storage.googleapis.com/my-bucket-name
```
Comment (jgeewax) [inline docs/logging-usage.rst]: is it worth having different classes for the different types of sinks ?

File: docs/logging-usage.rst (hunk 9 above)
Comment (tseaver) [inline docs/logging-usage.rst]: @jgeewax the issue with hiding the token/page in favor of a more convenient `__iter__` is that the expense (both time and actual charges incurred) is hidden, too.  With logging, one could literally be pulling down hundreds of thousands of log entries, inadvertently.

I would propose that we update our iterator such that users could do:

```python
from gcloud.iterator import Iterator

for entry in Iterator(client.list_entries):
    print(entry.name)
```

(Currently, the iterator implementation is based around a path, rather than a method which returns a list and a token).

File: docs/logging-usage.rst (hunk 33 above)
Comment (tseaver) [inline docs/logging-usage.rst]: I was thinking that the validation logic (for handling the only-one-target bit you point out below) would be easiest in a property setter.

File: docs/logging-usage.rst (hunk 34 above)
Comment (tseaver) [inline docs/logging-usage.rst]: The [`projects.sinks.destination` field](https://cloud.google.com/logging/docs/api/ref_v2beta1/rest/v2beta1/projects.sinks#LogSink.FIELDS.destination) is single-valued.  I meant to block the possibility of setting more than one in the property setters.

File: docs/logging-usage.rst (hunk 35 above)
Comment (tseaver) [inline docs/logging-usage.rst]: I don't think so:  they don't have any different methods.  The three properties are basically sugar for generating the appropriate `destination` URL.

File: docs/logging-usage.rst (hunk 30 above)
Comment (tseaver) [inline docs/logging-usage.rst]: 2e0169a

File: docs/logging-usage.rst (hunk 31 above)
Comment (tseaver) [inline docs/logging-usage.rst]: Good catch!  d973ca7

File: docs/logging-usage.rst (hunk 32 above)
Comment (tseaver) [inline docs/logging-usage.rst]: OK.  I was mapping the [`projects.logs.delete` API method](https://cloud.google.com/logging/docs/api/ref_v2beta1/rest/v2beta1/projects.logs/delete).  559e7cd

**Comment by tseaver** [comment]
@dhermes @jgeewax Any remaining issues?  I will merge manually to the `logging-api` branch (rather than to `master`), and have another PR queued up afterwards.

File: docs/logging-usage.rst (hunk 9 above)
Comment (tseaver) [inline docs/logging-usage.rst]: #1554 outlines my proposed the changes to `Iterator`.

**Comment by tseaver** [comment]
Squashed and pushed to the `logging-api` branch via d0c0bb0bd066be12c78b13e86f817acafc945912

**Comment by dhermes** [comment]
:+1:
